### ISSUE TITLE
Devices API

### ISSUE THREAD

**Issue by joshlk** [pytorch/pytorch#110080]
### 🚀 The feature, motivation and pitch

Hi,

I would like to propose a new device API that is device-type (cuda, xla, npu, …) agnostic. For example, you could do:
```
torch.devices.count(type='cuda')
torch.devices.is_available(type='npu')
torch.devices.seed(42, type='xla')
```

Instead of:
```
torch.cuda.device_count()
torch_npu.npu.is_available()
torch_xla.set_rng_state(42)
```

This would greatly help with making 3rd party code device-type agnostic because, at the moment, each device type has its own API.

I would be happy to implement such an API if there is support from the community.

#108046 is loossly related and could be put under the same API namespace

(cc team: @AnthonyBarbier @HMellor)

**Comment by albanD** [comment]
Note that an alternative is to clean up our "device modules" to be more consistent and allow users to be able to do:
```
d = torch.cpu
d = torch.cuda
d = torch_npu.npu

# Where the same API is available on all versions of d
d.is_built()
d.is_available()
etc
```

cc @FFFrog as well as we did this standardization work partially for the privateuse1 device.

**Comment by joshlk** [comment]
> Note that an alternative is to clean up our "device modules" to be more consistent and allow users to be able to do

Yes that could work. I would suspect it would be harder to ask many venders to clean up their own API's. We would also need a way to map between device type and API namespace. e.g.

```
d = torch.get_device_api('xla')
d.is_built()
d.is_available()
etc
```

It might be also useful to have a generic interface to work with any device type (similar to what #108046 proposed):
```
torch.devices.is_available()  # if any device type is available
devices = torch.devices.available(). # returns ['X'] (alphabetically ordered)
torch.zeros(10, device=devices[0])
```

**Comment by malfet** [comment]
We can not just deprecate an old APIs, nor is it beneficial to substantially increase the supported API surface that essentially does the same. 

Also, it looks like there is already a`torch.device()` which could logically be extended to support `is_available` or `get_rng_seed`/`set_rng_seed`  methods.

**Comment by albanD** [comment]
Note that this is only for inductor and doesn't necessarily match the "device" concept from our python API.

**Comment by rwightman** [comment]
@albanD @malfet support for non cuda/cpu devices has improved significantly in past few torch releases, but the issue this API refers to is still present. Can this be prioritized a bit? The autoload device extension is great, having autocast/amp functions that take device / device string args is great, but we still have an ugly mess of ifs, etc to check is_available, set device specific rng, use streams (if supported) etc.

It'd be very useful to have a common device api via torch.device() or as proposed here torch.devices.is_available(device), that can be used without torch.cuda.is_ .. torch.backends.mps.. or now numerous autoloaded extensions.

It would also be useful to query all devices that are available, and possibly have capabilities flags for things like streams that may be supported by many but not all devices, etc.

**Comment by FFFrog** [comment]
> @albanD @malfet support for non cuda/cpu devices has improved significantly in past few torch releases, but the issue this API refers to is still present. Can this be prioritized a bit? The autoload device extension is great, having autocast/amp functions that take device / device string args is great, but we still have an ugly mess of ifs, etc to check is_available, set device specific rng, use streams (if supported) etc.
> 
> 
> 
> It'd be very useful to have a common device api via torch.device() or as proposed here torch.devices.is_available(device), that can be used without torch.cuda.is_ .. torch.backends.mps.. or now numerous autoloaded extensions.
> 
> 
> 
> It would also be useful to query all devices that are available, and possibly have capabilities flags for things like streams that may be supported by many but not all devices, etc.
> 
> 

Hi, you can refer to this [pr](https://github.com/pytorch/pytorch/pull/132204) to get more info about device-agnostic
